KAFKA-2804: manage changelog topics through ZK in PartitionAssignor

guozhangwang opened the pull request.

Commit:
comment links

Commit:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K2804

guozhangwang committed:
auto create topic in partition assignor and block wait on topic partition existence

guozhangwang committed:
fix unit tests

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K2804

guozhangwang commented:
Ping @ymatsuda to take a look, I'm still adding some more unit tests but the basic logic is there.

ymatsuda commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -332,20 +334,31 @@ private void connectProcessorAndStateStore(String processorName, String stateSto
+                // if the node is connected to a state, add to the state topics
+                for (Map.Entry<String, Set<String>> stateUsers : stateStoreUsers.entrySet()) {
+                    if (stateUsers.getValue().contains(node)) {
+                        stateTopics.add(stateStores.get(stateUsers.getKey()).name());
```
I wonder how to handle a topic if a state store is used for a ktable. Should it be excluded from stateTopics?

guozhangwang replied:
Yes, it should be excluded from the state topics.

ymatsuda replied:
I may be missing some thing, but are ktable topics excluded?

ymatsuda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/KafkaStreamingPartitionAssignor.java:
```diff
@@ -183,6 +317,54 @@ public Subscription subscription(Set<String> topics) {
+
+            for (Map.Entry<String, Set<TaskId>> entry : tasksInfo.tasksForState.entrySet()) {
+                String topic = entry.getKey() + ProcessorStateManager.STATE_CHANGELOG_TOPIC_SUFFIX;
+                int numTasks = entry.getValue().size();
```
The number of change log partitions may not be equal to the number of tasks.
It is the max value of taskId.partition, isn't it?

guozhangwang replied:
That is right, will fix.

ymatsuda commented on streams/src/main/java/org/apache/kafka/streams/processor/PartitionGrouper.java:
```diff
@@ -26,19 +26,71 @@
+
+    public static class TasksInfo {
+        public Map<TaskId, Set<TopicPartition>> partitionsForTask;
+        public Map<String, Set<TaskId>> tasksForState;
```
I don't like that PartitionGrouper being concerned with state store change log topics. It seems possible to move the logic in DefaultPartitionGrouper to KafkaStreamingPartitionAssigner and to make PartitionGrouper only do grouping of partitions. PartitionGrouper something that a user can implement. It doesn't seem right to have the code related to change log partition creation there. It is not meant to be customized, right?

guozhangwang replied:
Yeah I agree, though the problem is that only TopologyBuilder knows the state topics to tasks mapping, and it is only accessible to PartitionGrouper of the StreamThread. 

I will change it to make the topology return only this map via StreamThread to KafkaStreamingPartitionAssignor.

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K2804

guozhangwang committed:
incoporate comments

guozhangwang committed:
refactor PartitionGrouper

guozhangwang committed:
add Job-Id into StreamingConfig

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K2804

guozhangwang committed:
some minor fixes

guozhangwang commented:
@ymatsuda Ping for another look. Some changes:
1. Added a Job Id config to StreamingConfig, which will be used as the group id and the default client id prefix if client id is not specified.
2. Override the consumer / producer / restore-consumer's client ids with some meaningful suffixes.
3. Removed ClientUUID, the reasoning is that client id could be set by users to make sure they are unique, if not we try to set it "in best effort" but does not guarantee uniqueness across JVMs.
4. The streaming client id is used in metrics as well as assignment member id as well.
5. Moved the change log validation as well as topics info into KafkaStreamingPartitionAssignor, and make PartitionGrouper light-weight.

guozhangwang committed:
merge from trunk and fix unit tests

guozhangwang commented:
Jenkins failures are not relevant, filed KAFKA-2943 to track it separately.

ymatsuda commented:
I want a guaranteed uniqueness across JVMs. I do not see why you want to remove clientUUID and use much less reliable clientId.

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K2804

guozhangwang committed:
revert changes to UUID

guozhangwang commented:
@ymatsuda Reverted changes to UUIDs.

ymatsuda commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -109,27 +111,36 @@ public void register(StateStore store, StateRestoreCallback stateRestoreCallback
-        if (restoreConsumer.listTopics().containsKey(store.name())) {
-            boolean partitionNotFound = true;
-            for (PartitionInfo partitionInfo : restoreConsumer.partitionsFor(store.name())) {
+        String topic = store.name() + STATE_CHANGELOG_TOPIC_SUFFIX;
```
Why adding a suffix? Is there any problem if the topic name is equal to the store name?
This breaks ktable.

guozhangwang replied:
I want to distinguish changelog topics with normal topics inside Kafka, this is mainly for operationability: otherwise users have to make sure their state name does not potentially conflict with any topics in a shared cluster, while it might be OK for them to make sure their state name is different with other state names from other jobs.

guozhangwang committed:
add job id as prefix of the change log topic

guozhangwang committed:
resolve conflicts merging in KTable

guozhangwang commented:
@ymatsuda Rebased on KTable.

ymatsuda commented on streams/src/main/java/org/apache/kafka/streams/processor/PartitionGrouper.java:
```diff
@@ -19,39 +19,18 @@
-    }
-
-}
+    public abstract Map<TaskId, Set<TopicPartition>> partitionGroups(Map<Integer, Set<String>> topicGroups, Cluster metadata);
```
Now PartitionGrouper has only one abstract method. Shall we make it an interface?

ymatsuda commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -332,20 +362,31 @@ private void connectProcessorAndStateStore(String processorName, String stateSto
         for (Map.Entry<Integer, Set<String>> entry : nodeGroups.entrySet()) {
-            Set<String> topicGroup = new HashSet<>();
+            Set<String> sourceTopics = new HashSet<>();
+            Set<String> stateTopics = new HashSet<>();
```
This set contains state names, not topic names. The topic name is no longer same as the state name.

guozhangwang committed:
make PartitionGrouper an interface

guozhangwang committed:
exclude KTable from internal state stores

guozhangwang committed:
Merge branch 'trunk' of https://git-wip-us.apache.org/repos/asf/kafka into K2804

guozhangwang committed:
use store name as topic name if logging is not enabled

guozhangwang commented:
@ymatsuda Add logic to exclude topics for KTables and use store name directly as topic name if loggingEnabled is false.

ymatsuda commented:
lgtm

gwenshap commented:
LGTM.

Can you create a JIRA to remove the ZK dependency after KIP-4 is done? So we won't leave temporary solutions in here?
